File: networking_benchmarks/af_xdp/control-plane/web/src/lib/pairs.js
// Target-set resolution, shared by the control panel, 2D and 3D.
//
// This mirrors backend/pairs.go. The backend is authoritative -- it re-resolves
// against its own Online() set before dispatching -- but the UI needs the same
// answer to label the run buttons and draw the pair overlay. The two are kept in
// step by identical test tables on both sides.

export const SCOPE_AMONG = 'among';
export const SCOPE_FANOUT = 'fanout';
export const SCOPE_FANIN = 'fanin';

export const SCOPES = [
  { id: SCOPE_AMONG, label: 'Among selected', hint: 'how these nodes see each other' },
  { id: SCOPE_FANOUT, label: 'Selected \u2192 all', hint: 'how these nodes reach the fleet' },
  { id: SCOPE_FANIN, label: 'All \u2192 selected', hint: 'how the fleet reaches these nodes (costly)' },
];

const idOf = (n) => n.instance_id || n.instanceId || n.private_ip || n.name;

/**
 * Expand a target set into the ordered pairs a campaign would measure.
 *
 * @param {Array} nodes  fleet nodes; only `online` ones are measurable
 * @param {Array<string>|Set<string>} ids  selected instance ids ([] = full mesh)
 * @param {string} scope  among | fanout | fanin
 * @returns {{pairs: Array<[object,object]>, sources: Array, skipped: Array<string>, error: string|null}}
 */
export function resolvePairs(nodes, ids, scope) {
  const want = [...(ids || [])];
  const sc = scope || SCOPE_AMONG;
  const up = (nodes || []).filter((n) => n.online);

  const empty = { pairs: [], sources: [], skipped: [], error: null };
  if (![SCOPE_AMONG, SCOPE_FANOUT, SCOPE_FANIN].includes(sc)) {
    return { ...empty, error: `unknown scope "${sc}"` };
  }

  // Resolve targets in fleet order so a run is deterministic regardless of the
  // order the user clicked, and de-duplicate repeated ids.
  const wanted = new Set(want);
  const targets = up.filter((n) => wanted.has(idOf(n)));
  const found = new Set(targets.map(idOf));
  const skipped = [...new Set(want.filter((id) => !found.has(id)))];

  // A caller that named specific nodes must never widen to the full mesh when
  // none of them resolve: that would run the entire fleet by accident.
  if (want.length > 0 && targets.length === 0) {
    return {
      ...empty,
      skipped,
      error: `none of the ${want.length} selected node(s) are online`,
    };
  }

  const exclSelf = (src, pool) => pool.filter((d) => idOf(d) !== idOf(src));

  let sources = [];
  const destsFor = new Map();
  if (want.length === 0) {
    if (up.length < 2) return { ...empty, skipped, error: `need >=2 online nodes, have ${up.length}` };
    sources = up;
    for (const s of up) destsFor.set(idOf(s), exclSelf(s, up));
  } else if (sc === SCOPE_AMONG) {
    sources = targets;
    for (const s of targets) destsFor.set(idOf(s), exclSelf(s, targets));
  } else if (sc === SCOPE_FANOUT) {
    sources = targets;
    for (const s of targets) destsFor.set(idOf(s), exclSelf(s, up));
  } else {
    // fanin: every online node measures to the targets. Sources with no dests
    // left (the node is the only target) are omitted.
    for (const s of up) {
      const d = exclSelf(s, targets);
      if (d.length) {
        sources.push(s);
        destsFor.set(idOf(s), d);
      }
    }
  }

  const pairs = [];
  for (const s of sources) for (const d of destsFor.get(idOf(s)) || []) pairs.push([s, d]);

  if (pairs.length === 0) {
    return {
      ...empty,
      skipped,
      error: `${scopeDescription(sc, targets.length)} resolves to 0 pairs`,
    };
  }
  return { pairs, sources, skipped, error: null };
}

/**
 * Pair count without a fleet snapshot, for button labels.
 *
 * fanout and fanin are both k*(N-1): a fanin source that is itself a target
 * contributes k-1 dests rather than k, and k*(N-k) + k*(k-1) reduces to k*(N-1).
 */
export function countPairs(n, k, scope) {
  if (n < 2) return 0;
  if (k === 0) return n * (n - 1); // full mesh; scope is ignored
  if (k > n) k = n;
  if (scope === SCOPE_FANOUT || scope === SCOPE_FANIN) return k * (n - 1);
  return k * (k - 1); // among
}

/** Human wording for the resolved scope. Matches backend scopeDescription. */
export function scopeDescription(scope, k) {
  if (k === 0) return 'full mesh';
  const w = k === 1 ? 'node' : 'nodes';
  if (scope === SCOPE_FANOUT) return `${k} selected ${w} to all`;
  if (scope === SCOPE_FANIN) return `all to ${k} selected ${w}`;
  return `among ${k} selected ${w}`;
}

/**
 * Drop target ids that are no longer online members of the fleet.
 *
 * Called on every re-render so a terminated or dropped-out node cannot keep
 * silently scoping runs. Returns a new Set; the input is not mutated.
 */
export function prunedTargets(targetIds, nodes) {
  const live = new Set((nodes || []).filter((n) => n.online).map(idOf));
  return new Set([...(targetIds || [])].filter((id) => live.has(id)));
}

File: networking_benchmarks/af_xdp/control-plane/backend/pairs/pairs.go
package pairs

import (
	"fmt"

	"afxdp-cp/backend/registry"
)

// Scope names accepted by ResolvePairs. They expand a target set into ordered
// pairs; an empty target set is a full mesh regardless of scope.
const (
	ScopeAmong  = "among"  // k*(k-1): how the selected nodes see each other
	ScopeFanout = "fanout" // k*(N-1): how the selected nodes reach the fleet
	ScopeFanin  = "fanin"  // (N-1)*k: how the fleet reaches the selected nodes
)

// ResolvePairs turns a target set + scope into the (sources, destsFor) shape the
// source-grouped campaign loop needs.
//
// Offline and unknown ids are dropped and returned in skipped so the caller can
// say so rather than silently measuring less than the user asked for. An empty
// target set means full mesh, which is the pre-existing behaviour.
func ResolvePairs(online []registry.Node, ids []string, scope string) (
	sources []registry.Node, destsFor map[string][]registry.Node, skipped []string, err error) {

	if scope == "" {
		scope = ScopeAmong
	}
	if scope != ScopeAmong && scope != ScopeFanout && scope != ScopeFanin {
		return nil, nil, nil, fmt.Errorf("unknown scope %q (want %s, %s or %s)",
			scope, ScopeAmong, ScopeFanout, ScopeFanin)
	}

	// Only online nodes are ever measurable. A node that dropped out between the
	// UI selecting it and the campaign starting must not be dispatched to.
	up := make([]registry.Node, 0, len(online))
	byID := map[string]registry.Node{}
	for _, n := range online {
		if !n.Online {
			continue
		}
		up = append(up, n)
		byID[n.InstanceID] = n
	}

	// Resolve the target set, preserving the fleet's own ordering so a run is
	// deterministic regardless of the order the user clicked nodes in.
	var targets []registry.Node
	if len(ids) > 0 {
		want := map[string]bool{}
		for _, id := range ids {
			want[id] = true
		}
		for _, n := range up {
			if want[n.InstanceID] {
				targets = append(targets, n)
				delete(want, n.InstanceID) // de-duplicate repeated ids
			}
		}
		for id := range want {
			skipped = append(skipped, id) // offline or not in the fleet at all
		}
	}

	destsFor = map[string][]registry.Node{}
	// exclSelf appends every node in pool except src.
	exclSelf := func(src registry.Node, pool []registry.Node) []registry.Node {
		out := make([]registry.Node, 0, len(pool))
		for _, d := range pool {
			if d.InstanceID != src.InstanceID {
				out = append(out, d)
			}
		}
		return out
	}

	// A caller that asked for specific nodes must never silently widen to the
	// full mesh when none of them resolve: that would run every pair in the
	// fleet instead of the handful requested.
	if len(ids) > 0 && len(targets) == 0 {
		return nil, nil, skipped, fmt.Errorf(
			"none of the %d requested node(s) are online: %v", len(ids), skipped)
	}

	switch {
	case len(ids) == 0:
		// Full mesh. Scope is meaningless here and deliberately ignored.
		if len(up) < 2 {
			return nil, nil, skipped, fmt.Errorf("need >=2 online nodes for a full mesh, have %d", len(up))
		}
		sources = up
		for _, s := range up {
			destsFor[s.InstanceID] = exclSelf(s, up)
		}
	case scope == ScopeAmong:
		sources = targets
		for _, s := range targets {
			destsFor[s.InstanceID] = exclSelf(s, targets)
		}
	case scope == ScopeFanout:
		sources = targets
		for _, s := range targets {
			destsFor[s.InstanceID] = exclSelf(s, up)
		}
	case scope == ScopeFanin:
		// Every online node measures to the targets. A source with no dests left
		// (it is the only target) is omitted so the loop does not pay a profile
		// transition for zero measurements.
		for _, s := range up {
			if d := exclSelf(s, targets); len(d) > 0 {
				sources = append(sources, s)
				destsFor[s.InstanceID] = d
			}
		}
	}

	n := 0
	for _, s := range sources {
		n += len(destsFor[s.InstanceID])
	}
	if n == 0 {
		return nil, nil, skipped, fmt.Errorf(
			"scope %q over %d selected node(s) resolves to 0 pairs", scope, len(targets))
	}
	return sources, destsFor, skipped, nil
}

// PrepareSet is the union of sources and every destination.
//
// The prepare phase converges hosts to the echo profile, and it must cover
// destinations too: a node left in client profile by an earlier run does not
// echo, so measuring to it would fail for a reason unrelated to the network.
func PrepareSet(sources []registry.Node, destsFor map[string][]registry.Node) []registry.Node {
	seen := map[string]bool{}
	var out []registry.Node
	add := func(n registry.Node) {
		if !seen[n.InstanceID] {
			seen[n.InstanceID] = true
			out = append(out, n)
		}
	}
	for _, s := range sources {
		add(s)
		for _, d := range destsFor[s.InstanceID] {
			add(d)
		}
	}
	return out
}

// ScopeDescription renders the resolved scope for the job event, so the UI log
// states what ran instead of only how many pairs it was.
func ScopeDescription(scope string, k int) string {
	if k == 0 {
		return "full mesh"
	}
	nodeWord := "nodes"
	if k == 1 {
		nodeWord = "node"
	}
	switch scope {
	case ScopeFanout:
		return fmt.Sprintf("%d selected %s to all", k, nodeWord)
	case ScopeFanin:
		return fmt.Sprintf("all to %d selected %s", k, nodeWord)
	default:
		return fmt.Sprintf("among %d selected %s", k, nodeWord)
	}
}

// CountPairs predicts the pair count without a fleet snapshot, for the UI button
// label. It mirrors ResolvePairs and returns 0 where that would resolve to
// nothing, so the label never promises a run that will be refused.
//
// fanout and fanin are both k*(N-1): a fanin source that is itself a target
// contributes k-1 dests rather than k, and k*(N-k) + k*(k-1) reduces to k*(N-1).
func CountPairs(n, k int, scope string) int {
	if n < 2 {
		return 0
	}
	if k == 0 {
		return n * (n - 1) // full mesh; scope is ignored
	}
	if k > n {
		k = n
	}
	switch scope {
	case ScopeFanout, ScopeFanin:
		return k * (n - 1)
	default: // among
		return k * (k - 1)
	}
}

// ScopeName is the scope as persisted in the runs table: "full" for a full mesh,
// otherwise the requested scope id. Distinct from ScopeDescription, which is
// prose for the UI log.
func ScopeName(scope string, k int) string {
	if k == 0 {
		return "full"
	}
	if scope == "" {
		return ScopeAmong
	}
	return scope
}
